ISO/IEC 19794-2:2005 Template IO

fatboab opened the pull request:
Import/Export methods for reading/writing ISO templates from SourceAFIS.

Commit:
Correct spellink...

Commit:
Start on ISO/IEC 19794-2:2005 template support.

Commit:
Initial stab at an ISO/IEC 19794-2:2005 template writer

Commit:
Add minutia to the ISO/IEC 19794-2:2005 template output.
Currently commented out as it core dumps.

Commit:
Stop the coredump when trying to free non dynamically allocated moemory...

Commit:
Switch back to removing from head, rather than tail.

Commit:
Add an ISO/IEC 19794-2:2005 template input routine...

Commit:
Merge branch 'master' into isotemplate-io

Commit:
Merge branch 'image-to-template-test' into isotemplate-io

fatboab commented:
Just realised we're not updating the template length in the file once we know it...

Commit:
Calculate and add in the template length when exporting

fatboab commented:
Good to be merged now...

andrew-farries commented:
The test `TemplateIO_ISO19794_2_2005` fails on my machine with: ` Expected 6 Was 500. originalDpi's are different`

It segfaults on @olane 's machine.

cjheppell commented:
FWIW, running this through `gdb` on OS X points to this on failure:
~~~
Program received signal SIGSEGV, Segmentation fault.
[..] in TemplateIO_ISO19794_2_2005_Export [..] at Templates/TemplateIO.c:117
117	        int16_t x = minutia->position.x;
~~~

Pointing to line 117 in `TemplateIO.c` as a possible issue?

Commit:
Merge branch 'master' into isotemplate-io

Commit:
Merge branch 'master' into isotemplate-io

Commit:
Merge branch 'image-to-template-test' into isotemplate-io

Commit:
Get rid of an auto-generated comment and some hanging chars on another...

Commit:
Include <stdlib.h> not <malloc.h>

Commit:
Merge branch 'image-to-template-test' into isotemplate-io

Commit:
Merge branch 'master' into isotemplate-io

Commit:
Merge branch 'fix-mem-leaks' into isotemplate-io

Commit:
remove the duplicate .c from the end fo the file name

fatboab commented:
Oh FFS! Turns out that on cygwin at least, the Unity TEST_SETUP stuff wasn't being called, so the template that was being written to disk was bogus. It's now failing on the Import portion of the test...

Commit:
Update the VS Solution

Commit:
Revert "Update the VS Solution"

This reverts commit 05110103aa72f983a61e03cc5a576255d7645b89.

Commit:
Merge branch 'master' into isotemplate-io

Commit:
Turns out the TEST_SETUP and TEST_TEAR_DOWN weren't being called when running cvia cygwin

Commit:
W00t! Fix the memory leakage by freeing the right stuff with the right free method...

Commit:
Merge branch 'master' into isotemplate-io

Commit:
Merge branch 'image-to-template-test' into isotemplate-io

fatboab commented:
@andrew-farries @cjheppell so @david-connell and I have fixed it to run under cygwin and fixed the resulting memory leaks. With the memory leak stuff that @fffej did, we hope this means this is now good to be merged.

fffej commented on Sources/Templates/Template.c:
```diff
@@ -4,7 +4,7 @@
 #include <stdlib.h>
 #include <assert.h>
 
-void Template_AddMinuitia(Template *template, TemplateMinutia *minutia)
```
Are we going to use `const` consistently?  I'd have thought both should be `const` in this situation?

fatboab replied:
I have no idea, we need to decide on a few things like that and then apply them to all the code in another branch I think...

fffej commented on Sources/Templates/TemplateIO.c:
```diff
@@ -0,0 +1,298 @@
+//      1B quality (ignored, zeroed)
+// 2B rubbish (extra data length, zeroed)
+// N*1B rubbish (extra data)
+void TemplateIO_ISO19794_2_2005_Export(Template *template, const char *outputFileName)
```
`const`

What about error handling?  Should this return non-zero on failure with an appropriate error code? (rather than `assert`?)

fatboab replied:
All the boiler plate code that in the project before we stared used assert, so we just continued. I suppose this is another of those library wide questions that we need to agree on...

david-connell replied:
I am happy with asserts in the test code, but not in the library, to me the library code should be bullet-proof no way it can go wrong. It can fail to load a file or save a file, but if it fails all is good nothing bad happens, bar the file not being saved.

(OK Yes I know tests aren't 2nd class citizens.)

fffej commented on Sources/Templates/TemplateIO.c:
```diff
@@ -0,0 +1,298 @@
+    fclose(output);
+}
+
+void TemplateIO_ISO19794_2_2005_Import(const char *inputFileName, Template *template) 
```
Ditto on the error handling.

fffej commented on Sources/Tests/EndToEnd/Test_EndToEnd.c:
```diff
@@ -72,7 +72,6 @@ static void ReadTemplate(const char *expectedFileName, Template *expectedTemplat
     assert(ret != EOF);
 }
 
-
```
I got really confused about the `UnityFree` stuff - can someone explain why it is necessary?

I see that `malloc` and `free` are redefined, but what's it doing differently that means we have to duplicate code in our tests?

fatboab replied:
@david-connell can explain it better than I can, but the Unity stuff adds extra stuff into the malloc so it can track it, if we free with the normal free, rather than the Unity free, then this extra stuff is leaked and the test fails.

david-connell replied:
Hi @fffej , the rule is pretty simple, if you allocate under unity then free under unity, if you malloc normally then free normally.
As @fatboab says it seems like Unity does it's own tracking, so you cann't free under Unity if it wasn't allocated under Unity. I am sure that there is ways of disabling this feature or turning it off or enhancing Unity, but I did not look at that...

fffej replied:
@david-connell Thanks!  I guess I'd love to disable that malloc stuff, since valgrind does that job for us.  

Most of the things I've mentioned are library related (`const`) or unrelated to this (use of unity), so I'm happy for this to be merged!
